integrated_channels/moodle/client.py

```diff
@@ -311,38 +312,191 @@ def _wrapped_create_course_completion(self, user_id, payload):
+            except (TypeError, IndexError):
+                self._delete_content_metadata(response[0].get('id'))
+                raise ClientError(
+                    'Moodle Client Content Metadata Creation failed to create post for course {} '
+                    'as forum could not be found'.format(
+                        response[0].get('shortname'),
+                    ),
+                    HTTPStatus.NOT_FOUND.value
+                )
+            except Exception as exc:
+                self._delete_content_metadata(response[0].get('id'))
+                raise ClientError(
+                    'Moodle Client Content Metadata Creation failed to create post for course {} '
+                    'due to exception: {}'.format(response[0].get('shortname'), str(exc)),
+                    HTTPStatus.INTERNAL_SERVER_ERROR.value
+                )
+            if post.json().get('warnings', None) or post.json().get('exception', None):
+                self._delete_content_metadata(response[0].get('id'))
+                raise ClientError(
+                    'Moodle Client Content Metadata Creation failed to create post for course {}'
+                    .format(response[0].get('shortname')),
+                    HTTPStatus.BAD_REQUEST.value
+                )
+            format_params = {
+                'wsfunction': 'core_course_update_courses',
+                'courses[0][id]': response[0].get('id'),
+                'courses[0][format]': course_format,
+            }
+            course_update_response = self._post(format_params)
+            if course_update_response.json().get('warnings', None) or \
+                course_update_response.json().get('exception', None):
+                self._delete_content_metadata(response[0].get('id'))
+                raise ClientError(
+                    'Moodle Client Course Creation failed to update course format for course {}. '
+                    'Changes rolled back. '.format(response[0].get('shortname')),
+                    HTTPStatus.BAD_REQUEST.value
+                )
+        else:
+            raise ClientError(
+                'Moodle Client Course Creation failed to create course {}'
+                .format(serialized_data['courses[0][shortname]']),
+                HTTPStatus.BAD_REQUEST.value
+            )
+
+        return 200, ''
+
+
     def update_content_metadata(self, serialized_data):
+        announcement = serialized_data.pop('courses[0][announcement]')
         moodle_course_id = self.get_course_id(serialized_data['courses[0][shortname]'])
         serialized_data['courses[0][id]'] = moodle_course_id
         serialized_data['wsfunction'] = 'core_course_update_courses'
-        return self._post(serialized_data)
 
-    @moodle_request_wrapper
+        response = self._wrapped_update_content_metadata(serialized_data)
+        response_text = json.loads(response.text)
+        if response_text.get('exception', None) or response_text.get('warnings', None):
```

Comment:
these Moodle responses pain me

Reply:
yeah. This _could_ be resolved in the `inner` wrapper we have as we already handle a lot of the weird types that get returned (such as list, str, int). However, I think because of some cases where we just pass `_post` all the way through, we need to have access to status codes and other response object attrs so that's why even when we go through the process of checking the body, we still return the whole response object and thus need to reparse this garbage. :/